src/deepsparse/server/config.py

```diff
@@ -12,152 +12,161 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-"""
-Configurations for serving models in the DeepSparse inference server
-"""
 
-import json
-import os
-from functools import lru_cache
-from typing import List
+from typing import Any, Dict, List, Optional, Tuple, Union
 
-import yaml
 from pydantic import BaseModel, Field
 
-from deepsparse import PipelineConfig
-from deepsparse.cpu import cpu_architecture
+from deepsparse import DEEPSPARSE_ENGINE, PipelineConfig
+from deepsparse.tasks import SupportedTasks
 
 
 __all__ = [
-    "ENV_DEEPSPARSE_SERVER_CONFIG",
-    "ENV_SINGLE_PREFIX",
     "ServerConfig",
+    "EndpointConfig",
+    "SequenceLengthsConfig",
+    "ImageSizesConfig",
 ]
 
+# these are stored as global variables instead of enum because in order
+# to save/load enums using yaml, you have to enable arbitrary code
+# execution.
+INTEGRATION_LOCAL = "local"
+INTEGRATION_SAGEMAKER = "sagemaker"
+INTEGRATIONS = [INTEGRATION_LOCAL, INTEGRATION_SAGEMAKER]
 
-ENV_DEEPSPARSE_SERVER_CONFIG = "DEEPSPARSE_SERVER_CONFIG"
-ENV_SINGLE_PREFIX = "DEEPSPARSE_SINGLE_MODEL:"
 
+class SequenceLengthsConfig(BaseModel):
+    sequence_lengths: List[int] = Field(
+        description="The sequence lengths the model should accept"
+    )
 
-class ServerConfig(BaseModel):
-    """
-    A configuration for serving models in the DeepSparse inference server
-    """
 
-    models: List[PipelineConfig] = Field(
-        default=[],
+class ImageSizesConfig(BaseModel):
+    image_sizes: List[Tuple[int, int]] = Field(
+        description="The list of image sizes the model should accept"
+    )
+
+
+class EndpointConfig(BaseModel):
+    name: Optional[str] = Field(
+        default=None,
         description=(
-            "The models to serve in the server defined by PipelineConfig objects"
+            "Name of the model used for logging & metric purposes. "
+            "If not specified 'endpoint-<index>' will be used."
         ),
     )
-    workers: str = Field(
-        default=max(1, cpu_architecture().num_available_physical_cores // 2),
+
+    route: Optional[str] = Field(
+        default=None,
+        description="Optional url to use for this endpoint. E.g. '/predict'. "
+        "If there are multiple endpoints, all routes must be specified. "
+        "If there is a single endpoint, '/predict' is default if not specified.",
+    )
+
+    task: str = Field(description="Task this endpoint performs")
+
+    model: str = Field(description="Location of the underlying model to use.")
+
+    batch_size: int = Field(
+        default=1, description="The batch size to compile the model for."
+    )
+
+    bucketing: Optional[Union[ImageSizesConfig, SequenceLengthsConfig]] = Field(
```

Comment:
nit: not `Union[Optional[ImageSizesConfig], Optional[SequenceLenghtsConfig]]?

Reply:
I prefer having the optional on the outside because the whole field is optional. Otherwise it could be valid to have a non-optional field in the union (e.g. `Union[Optional[ImageSizesConfig], SequenceLengthsConfig]`), which is not what the intention is.